### PR TITLE
Upgrade PHPUnit to ^9.0 and replace deprecated test methods

### LINKED ISSUE #397: Dependency upgrade to phpunit 9.0
Some of the methods we use in the tests are deprecated in phpunit 9. To sync with PHP 8, we got a dependabot alert to upgrade phpunit to 9.0, but we will need to find a way to replace deprecated methods first.

Related #396 
[AB#8213](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/8213)

### LINKED ISSUE #419: Upgrade minimum required PHP version
We currently support a minimum of `PHP 7.1` however this version of PHP is no longer supported (https://www.php.net/supported-versions.php)

`PHP 7.3` only has a couple of months of security support left (until 6 Dec 2021) hence the proposal to bump up our minimum version to `7.4`. 
[AB#8631](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/8631)

### PR DESCRIPTION
**Pull request by Ndiritu**
- Replaces deprecated methods with valid equivalents in v9.x
- Adds getters to access some private/protected variables' values in the tests
- Updates version string passed to setApiVersion() usage in tests to be consistent with ideal end-user usage (i.e. no backslash).
- Updates minimum required PHP version to PHP 7.4 

Closes #397 
Closes #419

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -17,7 +17,7 @@
 		"ext-json": "*"
 	},
 	"require-dev": {
-		"phpunit/phpunit": "^7.5 || ^8.0",
+		"phpunit/phpunit": "^9.0",
```
Comment (zengin) [inline composer.json:20]: just for sanity, does this upgrade work for both php versions (i.e. 8.0 and 7.1)?

File: tests/GraphTest.php [hunk 2]
```diff
@@ -44,18 +44,17 @@ public function testRequestWithCustomEndpoint()
...
     public function testBetaRequest()
     {
         $graph = new Graph();
         $graph->setAccessToken('abc')
-              ->setApiVersion('/beta');
+              ->setApiVersion('beta');
         $request = $graph->createRequest("GET", "/me");
 
-        $this->assertEquals('/beta', $this->readAttribute($request, 'apiVersion'));
+        $this->assertEquals('beta', $request->getApiVersion());
```
Comment (zengin) [inline tests/GraphTest.php:57]: Could you please confirm if not passing in `/` will create the correct URL using the HttpClient?

**Review by zengin — CHANGES_REQUESTED** [review]
waiting on confirmation of PHP versions and whether version requires a forward slash for building the URL.

File: tests/GraphTest.php (hunk 2 above)
Comment (MIchaelMainer) [reply tests/GraphTest.php:57]: To piggyback on this [comment ](https://github.com/microsoftgraph/msgraph-sdk-php/pull/418/files#r594530989)as a related note, we should get some clarity on how we handle leading `/` throughout the SDK. Ideally, the customer doesn't have to add it.

**Review by MIchaelMainer — COMMENTED** [review]
Looks good. Waiting to see response to Mustafa's comments.

File: tests/GraphTest.php (hunk 2 above)
Comment (Ndiritu) [reply tests/GraphTest.php:57]: I can confirm that it works without the leading `/`. Guzzle handles both cases (with or without the leading `/`) to generate the correct URL based on some [relative  URI resolution standard](https://tools.ietf.org/html/rfc3986#section-5.2). The [Guzzle `base_uri` docs](https://docs.guzzlephp.org/en/stable/quickstart.html) provide a good summary table

File: composer.json (hunk 1 above)
Comment (Ndiritu) [reply composer.json:20]: Thanks for calling this out.  PHPUnit needs a minimum of `7.3`. Also, `php 7.1` is no longer supported. Thinking we should consider upgrading to a minimum of `php 7.4`.

Comment (Ndiritu) [reply composer.json:20]: Added a commit to bump this up.

File: tests/GraphTest.php (hunk 2 above)
Comment (zengin) [reply tests/GraphTest.php:57]: That's awesome. Thanks for the confirmation!

File: composer.json (hunk 1 above)
Comment (zengin) [reply composer.json:20]: Unfortunately 7.3 to 7.4 is not a backwards compatible upgrade: https://www.php.net/manual/en/migration74.incompatible.php

I am assuming 7.1 to 7.4 will have more incompatibilities. @MIchaelMainer, do we have any community feedback regarding which version we should be on?

@Ndiritu alternatively, could we do something like: `^8.0 || ^9.0` for phpunit?

Comment (Ndiritu) [reply composer.json]: @zengin , makes sense. Added the option for phpunit ^8.0. However this needs a minimum of php 7.2. I've updated the minimum php version for the SDK to php 7.2 for now. Confirmed that unit tests pass.

Comment (zengin) [reply composer.json]: @Ndiritu thanks for making the changes. There are still incompatibilities between 7.1 and 7.2: https://www.php.net/manual/en/migration72.incompatible.php

Given that it is a relatively older minor version we are upgrading to, I am assuming that most customers will already be there. We need at least a minor version bump for this.

To me, any breaking change needs to go out with a major bump according to [SemVer](https://semver.org/). Here is a nice bit on that topic: https://semver.org/#if-even-the-tiniest-backwards-incompatible-changes-to-the-public-api-require-a-major-version-bump-wont-i-end-up-at-version-4200-very-rapidly

That said, I think we can align with PHP's philosophy here which allows minor breaking changes. @MIchaelMainer, what do you think?

Comment (Ndiritu) [reply composer.json]: @zengin Thanks for highlighting incompatibilities between PHP 7.1 and 7.2

More thoughts... @MIchaelMainer @zengin @SilasKenneth @roinochieng 
Was looking at the most recent PHP version usage metrics on Packagist [here](https://blog.packagist.com/php-versions-stats-2020-1-edition/). Given that most packages still cover php 7.1 and majority usage is only at php 7.3. I'm thinking the impact of this change to 7.2 should warrant a major version upgrade.

Comment (MIchaelMainer) [reply composer.json]: We should move forward with the LTS versions of PHP and target minimum at 7.3, with plans for targeting 7.4 when 7.3 is no longer in LTS.

If a customer wants to stay with 7.1, which is out of LTS, here are the options they have:

1. Upgrade to 7.3 and do the work to support it in the case there are incompatibilities. 
2. In general, the most likely APIs to be adopted are the ones that have been implemented earlier. So, our older SDK versions will likely cover most integration scenarios encountered by our customers. We should keep our eyes out for scenarios where customers are stuck in an older PHP version but need new versions of generated models - if we encounter  this, then we can reevaluate.
3. In the case that a customer is pegged to 7.1, and needs to use a new model not generated for a 7.1 version of the client library, we have a mechanism for them to use the raw request and response. This is a supported scenario and was implemented for this case. They could implement their own model as well.

Comment (MIchaelMainer) [reply composer.json]: @zengin The only community feedback I recall is that we should target the latest supported versions of PHP.

Comment (Ndiritu) [reply composer.json]: @zengin Upgraded minimum php version to 7.3 and bumped up the SDK minimum version

**Review by zengin — APPROVED** [review]
LGTM.

@Ndiritu thanks for your due diligence.

**Review by MIchaelMainer — APPROVED** [review]

**Review by SilasKenneth — APPROVED** [review]

**Pull request merged**